src/lib/panorama.ts

```diff
@@ -259,147 +267,187 @@ function getTileUrl(pano: Panorama, x: number, y: number, zoom: number) {
 	}
 }
 
-type PanoMetadata = NonNullable<Awaited<ReturnType<typeof getPanoMetadataFromId>>>;
-export function renderPanoFromMetadata(
+const envCanvas = document.createElement('canvas');
+const envContext = envCanvas.getContext('2d')!;
+
+const envTexture = new THREE.CanvasTexture(envCanvas);
+envTexture.mapping = THREE.EquirectangularReflectionMapping;
+envTexture.colorSpace = THREE.SRGBColorSpace;
+
+function setSceneEnvironment(
 	meta: PanoMetadata,
 	scene: THREE.Scene,
 	pmremGenerator: THREE.PMREMGenerator,
 	heading: number,
-	maxZoom: number,
 ) {
-	const zoom = Math.min(meta.maxZoom, maxZoom);
-
-	const sliceHorizontal = TAU / (meta.zoomLevels[zoom].cropWidth / meta.tileWidth);
-	const sliceVertical = Math.PI / (meta.zoomLevels[zoom].cropHeight / meta.tileHeight);
-
-	const panorama = new THREE.Group();
-	panorama.name = 'panorama';
-	panorama.rotation.order = 'YZX';
-	panorama.rotation.set(
-		THREE.MathUtils.degToRad((360 + meta.roll) % 360),
-		THREE.MathUtils.degToRad((360 + (90 - meta.heading) + heading) % 360),
-		THREE.MathUtils.degToRad((360 + (meta.tilt - 90)) % 360),
-	);
+	// TODO: Allow different environment quality levels (requires stitching multiple tiles together)
 
-	const loadingManager = new THREE.LoadingManager();
-	loadingManager.onLoad = () => {
-		newPano = panorama;
-		scene.add(panorama);
-		panoFadeTime = performance.now();
-		setSceneEnvironment(meta.pano, scene, pmremGenerator, meta.zoomLevels[0].cropWidth, meta.zoomLevels[0].cropHeight);
-	};
-	loadingManager.onError = (url) => {
-		console.error(`Couldn't load tile: ${url}`);
-		// renderErrorPano(scene, meta.panoId);
-	};
+	// The environment doesn't need to be high quality,
+	// so get the entire panorama as a single tile by using zoom level 0
+	const { cropWidth, cropHeight } = meta.zoomLevels[0];
 
-	for (let y = 0; y < meta.zoomLevels[zoom].numTilesY; y++) {
-		for (let x = 0; x < meta.zoomLevels[zoom].numTilesX; x++) {
-			const startX = sliceHorizontal * x;
-			let widthX = sliceHorizontal;
-			if (startX + widthX > 2 * Math.PI) widthX = 2 * Math.PI - startX;
+	envCanvas.width = cropWidth;
+	envCanvas.height = cropHeight;
 
-			const startY = sliceVertical * y;
-			let widthY = sliceVertical;
-			if (startY + widthY > Math.PI) widthY = Math.PI - startY;
+	// Get amount of pixels to shift by to simulate a yaw turn according to meta.heading
+	// so that the environment actually faces the direction we're driving.
+	const shift = Math.floor((((360 + (90 - meta.heading) + heading) % 360) / 360) * cropWidth);
```

Comment:
Remind me why we can't use [`scene.environmentRotation`](https://threejs.org/docs/#api/en/scenes/Scene.environmentRotation)?